awscli/autocomplete/serverside/servercomp.py

```diff
@@ -0,0 +1,148 @@
+        return lazy_call(self._session_import_name)
+
+
+class ServerSideCompleter(BaseCompleter):
+    def __init__(self, completion_lookup, client_creator):
+        self._completion_lookup = completion_lookup
+        self._client_creator = client_creator
+
+    def complete(self, parsed):
+        """
+
+        :param parsed: The parsed CLI coommand the user has typed.
+        :type parsed:  awscli.autocomplete.parser.ParsedResult
+        :return: A list of completion results, or None.
+        :rtype: List[awscli.autocomplete.completer.CompletionResult]
+
+        """
+        if parsed.unparsed_items or parsed.current_fragment is None:
+            return
+        # This completer only applies to option values.  If the current fragment
+        # we're on isn't an option value we can short circuit and return.
+        if not self._on_cli_option_value_fragment(parsed):
+            return
+        completion_data = self._completion_lookup.get_server_completion_data(
+            parsed.lineage, parsed.current_command, parsed.current_param,
+        )
+        if completion_data is None:
+            return
+        # At this point we know that this completer matches and we have
+        # the information we need to actually generate completion suggestions.
+        # We also know that we're only generate completoins that have a single
+        # list element.  We'll need to update this once we support completions
+        # with a 'parameters' key.
+        raw_results = self._retrieve_remote_completion_data(
+            parsed, completion_data['completions'][0])
+        return self._convert_to_completion_data(raw_results, parsed)
+
+    def _convert_to_completion_data(self, raw_results, parsed):
+        index = -len(parsed.current_fragment)
+        return [CompletionResult(r, index) for r in raw_results
+                if r.startswith(parsed.current_fragment)]
+
+    def _on_cli_option_value_fragment(self, parsed):
+        if parsed.current_param is None or parsed.current_fragment is None:
+            return False
+        # We're only handling the direct 1-1 mapped commands.  We could
+        # theoretically, but don't currently, handle modeled commands such
+        # as waiters.
+        if len(parsed.lineage) != 2:
+            return False
+        return True
+
+    def _retrieve_remote_completion_data(self, parsed, completion_data):
+        # TODO: Handle global params that can affect how we create the client
+        # (region, endpoint-url, profile, timeouts, etc).
+        service_name = self._get_service_name_from_command(parsed)
+        client = self._client_creator.create_client(service_name)
+        py_name = xform_name(completion_data['operation'])
+        api_params = self._map_command_to_api_params(parsed, completion_data)
+        response = self._invoke_api(client, py_name, api_params)
+        if response:
+            return jmespath.search(completion_data['jp_expr'], response)
```

Comment:
The only way that would happen is if the top level key in the response was missing, for example: `{}` instead of `{"Instances": []}`.  I'm not aware of that ever happening in a response but I suppose it's technically possible.  I'll update the code to handle this.